setup.py

```diff
@@ -1,9 +1,10 @@
 import os
```

Comment:
The `setup.py` is not used anymore, `pdm` reads all the information from `pyproject.toml` file. This is confusing and difficult to maintain since both files should be in sync.

Reply:
If you are suggesting to remove setup.py, I agree.

Reply:
setup.py is referenced in 5 different places of the repo. I will resolve them.

Reply:
c62dcb5d48ceea3414d59bb85faffbd53b1f1a2f removes setup.py and refactors github workflows to use pdm. This should be reviewed and tested by someone other than myself before merging.

Reply:
Additionally, I noticed multiple statements of:

```sh
pip install docker
pip install qdrant_client[fastembed]
```

but `docker` nor `qdrant_client` are specified as dependencies anywhere. We should add them. I'm guessing they would belong in the production deps? Do they need their own dependency group(s)?

I think `docker` should probably be in the main dependency list. I'm just curious why it's not already there, so thinking there must have been a reason for it, have not yet included it.

Reply:
I am not sure about the `docker` and `qdrant_client` deps. @sonichi, @ekzhu can you please provide some insight?

Reply:
@bitnom can you please resolve the conflict with `main`?

Reply:
Docker has previously been optional. If it's there, we use it. If it's not there, we do native code execution. That is in the process of being changed (in which case it should be added to the general dependencies).

Reply:
The current trend is to keep the main dependencies to the absolute minimum and to keep the rest as optional dependencies. Many of those optional dependencies are needed for testing or for running examples in notebooks. For testing, they should be grouped under the "test" optional target. For notebooks, I think it is best to declare those dependencies in a notebook itself. As the number of notebooks grows, they will eventually be in conflict anyway.

Reply:
Yes that is what I was referring to: https://github.com/microsoft/autogen/blob/b548e553b606a6efbcb5e0c4d2f786939d313765/setup.py#L41-L58

Reply:
I think we should remove optional targets specific for running notebooks or examples such as "blendsearch" and "mathchat" and leave only "test".

Reply:
That's a good idea. It reduces the complexity of the installation process. It's better to include them separately either in respective notebooks or custom modules, like how we do it for /sample/apps

Reply:
> I think we should remove optional targets specific for running notebooks or examples such as "blendsearch" and "mathchat" and leave only "test".

It's really no problem to keep the optional dependency groups in pyproject.toml, as they are low priority deps which we can later remove with no consequence. Otherwise, we will inevitably begin to see issues posted due to people struggling to run the demos.

I rather like the idea of having all these demos' deps pinned in optional groups as it helps keep the repo compatible with the deps all the cool kids are wanting to pin.

Reading through this just made me realize the `test` group deps should probably instead just be in `[tool.pdm.dev-dependencies]`

A related note for anyone coming upon this PR not having used the `pdm install` command before:
```
Dependencies Selection:
  -G GROUP, --group GROUP
                        Select group of optional-dependencies separated by comma or dev-dependencies (with `-d`). Can be supplied multiple times, use ":all" to include all
                        groups under the same species.
  --no-default          Don't include dependencies from the default group
  -d, --dev             Select dev dependencies
  --prod, --production  Unselect dev dependencie
```

Reply:
We have a problem with matplotlib version not being available in Python 3.8 because of an example in a notebook. Can we still support Python 3.8 if we have matplotlib in an optional dependency?

Reply:
> We have a problem with matplotlib version not being available in Python 3.8 because of an example in a notebook. Can we still support Python 3.8 if we have matplotlib in an optional dependency?

maybe by pinning an older version of matplotlib but I haven't tried it yet. A separate update regarding that: https://github.com/microsoft/autogen/discussions/1173#discussioncomment-8086095

where I change my position about dropping 3.8 support.

Reply:
#1167 has been resolved.